flac.go

```diff
@@ -253,7 +278,99 @@ func (stream *Stream) ParseNext() (f *frame.Frame, err error) {
 	return frame.Parse(stream.r)
 }
 
-// Seek is not implement yet.
-func (stream *Stream) Seek(offset int64, whence int) (read int64, err error) {
-	return 0, nil
+// Seek to offset where offset represents the sample number in the flac file
+//
+// TODO: return an error if sample is invalid?  Do nothing?
+//
+// sample could be invalid if:
+// whence == io.SeekEnd && sample is not negative
+// whence == io.SeekStart && sample is negative
+// whence == io.SeekCurrent and sample + current sample < 0 or > stream.Info.NSamples
+func (stream *Stream) Seek(sample int64, whence int) (read int64, err error) {
+	if stream.seekTable == nil {
+		return 0, nil
+	}
+
+	r := stream.r.(io.ReadSeeker)
+
+	var point meta.SeekPoint
+	switch whence {
+	case io.SeekStart:
+		point = stream.searchFromStart(sample)
+	case io.SeekCurrent:
+		o, err := r.Seek(0, io.SeekCurrent)
+		if err != nil {
+			return 0, err
+		}
+		point = stream.searchFromCurrent(sample, o)
+	case io.SeekEnd:
+		point = stream.searchFromEnd(sample)
+	default:
+		return 0, fmt.Errorf("invalid whence %d, must one of %d, %d or %d", whence, io.SeekStart, io.SeekCurrent, io.SeekEnd)
+	}
+
+	_, err = r.Seek(stream.start+int64(point.Offset), io.SeekStart)
+	return int64(point.SampleNum), err
+}
+
+func (stream *Stream) searchFromCurrent(sample, offset int64) (p meta.SeekPoint) {
+	for _, sp := range stream.seekTable.Points {
+		if int64(sp.Offset) >= offset {
+			return stream.searchFromStart(int64(sp.SampleNum) + sample)
+		}
+	}
+	return p
+}
+
+func (stream *Stream) searchFromEnd(sample int64) (p meta.SeekPoint) {
+	p = stream.seekTable.Points[len(stream.seekTable.Points)-1]
+	return stream.searchFromStart(int64(p.SampleNum) + sample)
+}
+
+func (stream *Stream) searchFromStart(sample int64) (p meta.SeekPoint) {
+	for _, p = range stream.seekTable.Points {
+		if p.SampleNum >= uint64(sample) {
+			return p
+		}
+	}
+	return p
+}
+
+func (stream *Stream) makeSeekTable() (err error) {
+	r := stream.r.(io.ReadSeeker)
+
+	var i int
+	var sampleNum uint64
+	var points []meta.SeekPoint
+	for {
+		f, err := stream.ParseNext()
+		if err == io.EOF {
+			break
+		}
+
+		if err != nil {
+			return err
+		}
+
+		if i%10 == 0 {
+			o, err := r.Seek(0, io.SeekCurrent)
```

Comment:
question: Second time I see this pattern, wondering if it would be better to create a function called something like `currentOffset`?

Reply:
Sure, no problem if you really want this.   However wouldn't that func have to do another stream.r.(io.ReadSeeker) assertion? So it'd either look like:

`func (stream *Stream) currentOffset() (int64, error) {
	rs, ok := stream.r.(io.ReadSeeker)
	if !ok {
		return 0, ErrNoSeeker
	}
	return rs.Seek(0, io.SeekCurrent)
}
`

Or

`func (stream *Stream) currentOffset(rs io.ReadSeeker) (int64, error) {
	return rs.Seek(0, io.SeekCurrent)
}
`

Reply:
I kind of like the way it is compared to either of those funcs, but am happy to make the change if you want.

Reply:
Yeah, it doesn't look that much better. Keep it as it is!